python/traverse.py

```diff
@@ -0,0 +1,227 @@
+
+ElementReplacement = Optional[
+    Union[str, lxml.html.HtmlElement, List[lxml.html.HtmlElement]]
+]
+
+
+def serialize_str_or_element(e) -> str:
+    if isinstance(e, str):
+        return e
+    return lxml.html.tostring(e, encoding="unicode")
+
+
+def append_text(element: lxml.html.HtmlElement, text: str) -> None:
+    stripped_text = text.lstrip()
+
+    if element.text:
+        print("appending element text", text)
+        stripped_element_text = element.text.rstrip()
+        sep = " " if stripped_element_text and stripped_text else ""
+        element.text = stripped_element_text + sep + stripped_text
+    else:
+        print("setting element text", text)
+        element.text = stripped_text
+
+
+def append_tail(element: lxml.html.HtmlElement, text: str) -> None:
+    stripped_text = text.lstrip()
+
+    if element.tail:
+        print("appending element tail", text)
+        stripped_element_tail = element.tail.rstrip()
+        sep = " " if stripped_element_tail and stripped_text else ""
+        element.tail = stripped_element_tail + sep + stripped_text
+    else:
+        print("setting element tail", text)
+        element.tail = " " + stripped_text
+
+
+def traverse_and_execute(
+    html: str, fn: Callable[[lxml.html.HtmlElement], None]
+) -> list:
+    elements = lxml.html.fragments_fromstring(html)
+
+    # Special case: just a string.
+    if len(elements) == 1 and isinstance(elements[0], str):
+        return elements
+
+    # If there's leading text, the first element of the array will be a string.
+    # We'll strip that off and add it back at the end, as we can't treat it
+    # like an element.
+    # TODO: update above comment.
+    if isinstance(elements[0], str):
+        del elements[0]
+
+    root_parent = elements[0].getparent()
+    children = root_parent.getchildren()
+    print("traversing into element", root_parent)
+    for element in children:
+        traverse_and_execute_impl(element, fn)
+
+    print("root_parent", root_parent, lxml.html.tostring(root_parent))
+    print("root_parent.getchildren()", root_parent.getchildren())
+
+    result_elements = root_parent.getchildren()
+    if root_parent.text:
+        result_elements.insert(0, root_parent.text)
+
+    print("result_elements", result_elements)
+    return result_elements
+
+
+def traverse_and_execute_impl(
+    element: lxml.html.HtmlElement,
+    fn: Callable[[lxml.html.HtmlElement], Union[None, List[lxml.html.HtmlElement]]],
+) -> None:
+    new_element = fn(element)
+    print("new_element", new_element)
+    has_new_element = new_element is not element and (
+        new_element is not None
+        and len(new_element) > 0
+        and new_element[0] is not element
+    )
+    traverse_element = (
+        new_element if has_new_element and new_element is not None else element
+    )
+    for child in traverse_element:
+        if not isinstance(child, str):
+            traverse_and_execute_impl(child, fn)
+
+
+def traverse_and_replace(
```

Comment:
Managed to write an iterative version of this that clocks in at about half the lines, I think it's conceptually simpler as well:
```python
def get_source_definition(element):
    #TODO reconstruct including attributes somehow
    return f"<{element.tag}>"


def traverse_and_replace(
    html: str,
    replace: Callable[
        [lxml.html.HtmlElement],
        ElementReplacement
    ]) -> str:
    import copy

    # Initialize result and work data structures
    result: List[Union[str, lxml.html.HtmlElement]] = []

    initial_list = lxml.html.fragments_fromstring(html)
    count_stack = [len(initial_list)]
    work_stack: List[Union[str, lxml.html.HtmlElement]] = copy.copy(initial_list[::-1])
    tail_stack = []

    while work_stack:
        element = work_stack.pop()

        # For just a string, append to final result
        if isinstance(element, str):
            result.append(element)

        else:
            new_elements = replace(element)

            # Turn new_elements into a list containing we can process
            if new_elements is None:
                new_elements = []
            elif isinstance(new_elements, str):
                fragments = lxml.html.fragments_fromstring(new_elements)
                new_elements = fragments if fragments is not None else []

            if isinstance(new_elements, list):
                # Modify count stack for new elements and decrement for element that was replaced
                count_stack[-1] += len(new_elements)-1

                # Add element tail before processing replaced element
                if element.tail is not None:
                    count_stack[-1] += 1
                    work_stack.append(element.tail)

                # Extend and go to the next iteration
                work_stack.extend(reversed(new_elements))
                continue

            # Add opening tag and text
            result.append(get_source_definition(new_elements))
            if new_elements.text is not None:
                result.append(new_elements.text)

            # Add all children to the work stack
            children = new_elements.getchildren()

            work_stack.extend(reversed(children))
            count_stack.append(len(children))
            tail_stack.append((new_elements.tag, element.tail))

        # Close all closable tags
        while count_stack[-1] == 0:
            count_stack.pop()
            tail_tag, tail_text = tail_stack.pop()

            result.append(f"</{tail_tag}>")
            if tail_text is not None:
                result.append(tail_text)
        else:
            count_stack[-1] -= 1

    # Empty out rest of tail tags.
    while tail_stack:
        tail_tag, tail_text = tail_stack.pop()

        result.append(f"</{tail_tag}>")
        if tail_text is not None:
            result.append(tail_text)

    return "".join(result)
```
@nwalters512 Passes all but one of the test cases (there was a convention for extra whitespace in the one that fails, wasn't sure how that should be so I chaged the test case locally). Haven't tested / benchmarked beyond the unit tests but seemed to be somewhat faster.

Reply:
Nice! I added a small benchmark:

```python
if __name__ == "__main__":
    html = """
    <pl-question-panel>
    <p> Consider two numbers $a = {{params.a}}$ and $b = {{params.b}}$.</p>
    <p> What is the sum $c = a + b$?</p>
    </pl-question-panel>

    <pl-integer-input answers-name="c" label="$c=$"></pl-integer-input>

    <pl-answer-panel>
      <pl-code-tabs filename="foo.py"></pl-code-tabs>
    </pl-answer-panel>
    """

    def replace(e: lxml.html.Element) -> str:
        if e.tag == "<pl-integer-input>":
            return "<pl-number-input></pl-number-input>"
        if e.tag == "pl-number-input":
            return '<input type="number" />'
        if e.tag == "pl-code-tabs":
            return "<div><pl-code></pl-code><pl-code></pl-code></div>"
        if e.tag == "pl-code":
            return "<pre><code>foo</code></pre>"

    original = timeit.timeit(lambda: traverse_and_replace(html, replace), number=1000)

    print("original", original)

    new = timeit.timeit(lambda: traverse_and_replace_new(html, replace), number=1000)

    print("new", new)
```

Your new version is pretty consistently faster:

```
original 0.0625948749948293
new 0.020077666005818173
```

Reply:
@nwalters512 great to see! I actually had some trouble getting consistent benchmarks for this because I wrote it on a really slow laptop 😅

Reply:
This was ultimately integrated into this PR ✅